File: dawn_python/xfast_trie.py
class XFastTrie:
    def __init__(self, w):
        u = 1 << w
        self.bits = [False] * (2 * u - 1)
        self.ones = []
        self.size = 0

    def insert(self, x):
        self._insert_aux(x, 0)

    def _insert_aux(self, x, i):
        if i >= len(self.bits):
            self.size += 1
            return
        bit = (x & (1 << i)) != 0
        index = 1 << (len(self.bits) - 1 - i)
        self.bits[index - 1] = self.bits[index - 1] or bit
        if bit:
            self.ones.append(index)
            self._insert_aux(x, i + 1)
        else:
            self._insert_aux(x, i + 1)

    def predecessor(self, x):
        return self._predecessor_aux(x, 0)

    def _predecessor_aux(self, x, i):
        if i >= len(self.bits):
            return None
        index = 1 << (len(self.bits) - 1 - i)
        if self.bits[index - 1]:
            left_child = 2 * index
            right_child = left_child + 1
            if self.bits[right_child - 1]:
                return self._predecessor_aux(x, i + 1)
            else:
                return self.ones[0]
        else:
            return self._predecessor_aux(x, i + 1)

File: dawn_cpp/xfast_trie.cpp
#include <vector>

class XFastTrie {
private:
    std::vector<bool> bits;
    std::vector<int> ones;
    int size;

public:
    XFastTrie(int w) {
        int u = 1 << w;
        bits = std::vector<bool>(2 * u - 1, false);
        ones = std::vector<int>();
        size = 0;
    }

    void insert(int x) {
        insertAux(x, 0);
    }

    int predecessor(int x) {
        int pred = predecessorAux(x, 0);
        if (pred == -1) {
            return -1;
        }
        while (pred > 0 && !std::binary_search(ones.begin(), ones.end(), pred)) {
            pred /= 2;
        }
        return pred - 1;
    }

private:
    void insertAux(int x, int i) {
        if (i >= bits.size()) {
            size++;
            return;
        }
        bool bit = (x & (1 << i)) != 0;
        int index = 1 << (bits.size() - 1 - i);
        bits[index - 1] = bits[index - 1] || bit;
        if (bit) {
            ones.push_back(index);
            insertAux(x, i + 1);
        } else {
            insertAux(x, i + 1);
        }
    }

    int predecessorAux(int x, int i) {
        if (i >= bits.size()) {
            return -1;
        }
        int index = 1 << (bits.size() - 1 - i);
        if (bits[index - 1]) {
            int leftChild = 2 * index;
            int rightChild = leftChild + 1;
            if (bits[rightChild - 1]) {
                return predecessorAux(x, i + 1);
            } else {
                return index;
            }
        } else {
            return predecessorAux(x, i + 1);
        }
    }
};